ldm/src/pqcat/pqcat.1
." $Id: pqcat.1,v 1.14.16.1.2.2 2009/06/18 16:15:03 steve Exp $
.TH PQCAT 1 "$Date: 2009/06/18 16:15:03 $"
.SH NAME
pqcat - program to print products from an LDM product queue
.SH SYNOPSIS
.HP
.ft B
pqcat
.nh
\%[-v]
\%[-O]
\%[-x]
\%[-l\ \fIlogfile\fP]
\%[-f\ \fIfeedtype\fP]
\%[-p\ \fIpattern\fP]
\%[-q\ \fIpqfname\fP]
\%[-i\ \fIinterval\fP]
\%[-o\ \fIoffset\fP]
\%[-c]
\%[-s]
\%[\fIoutputfile\fP]
.hy
.ft
.SH DESCRIPTION
.LP
This program writes to \fIoutputfile\fP selected data products from a local
LDM product queue (see \fBpq\fP(3)). If no \fIoutputfile\fP is specified,
the data is written to standard output.  The program may also be used to log
product information about the products in a product queue whose identifiers
match a specified feedtype and pattern.  By default,
.B pqcat
starts at the front of the product queue (the oldest products) and iterates
through products in order until it reaches the end of the queue (the most
recently inserted products).  On reaching the end of the queue it exits,
unless a non zero \fIinterval\fP is specified.
.LP
.B pqcat
is typically used to see what is in the product queue or to select
particular products out of the product queue.
.SH OPTIONS
.TP
.B -v
Verbose logging.  A line is emitted for every product in the queue whose
feed type and identifier match the specified feedtype and pattern.  The
emitted line contains the UTC date and time, the program name, the product
ingest time, feedtype, sequence number, size in bytes, and the product
identifier.
.TP
.B -O
Show product origin.  Adds originating site of product to each line of
verbose output.  Valid only with -v option.
.TP
.B -x
Debugging information is also emitted.
.TP
.BI "-l " logfile
The path name of a file to be used as the log file for the process.  The
default is to use standard error when interactive and syslogd(8) otherwise.
To use syslogd from the command line, enter ``pqcat ... >& /dev/null''.

.TP
.BI \-f " feedtype"
Reads from the product queue only products that have a feedtype that is a
member of the \fIfeedtype\fP set.  The default is `ANY', specifying all
feed types.  See \fBpqact\fP(1) for a list of basic feed types
and the use of feedtype expressions to specify combinations of basic
feed types.
.TP
.BI \-p " pattern"
Reads from the product queue only products whose identifier
matches the regular expression \fIpattern\fP.
The default is `\fB.*\fP', specifying all products.
.TP
.BI "-q " pqfname
The filename of the product queue.
The default is \fBdata/ldm.pq\fP relative to the installation point,
which is typically \fB/usr/local/ldm\fP.
.TP
.BI \-i " interval"
Polling interval, in seconds.  When the end of the queue is reached, the
program sleeps and
checks for new products in the product queue every \fIinterval\fP seconds.
If the \fIinterval\fP is 0, the program exits after one pass through the queue.
When
.B pqcat
is run in the same process group as the programs that insert products into
the product queue, a signal informs
.B pqcat
and all other interested processes in the process group
whenever a new product is available. This may wake up the process sooner than
\fIinterval\fP.
.TP
.BI \-o " offset"
Offset time, in seconds.
Begin reading products inserted into the product queue \fIoffset\fP
seconds earlier than the current time.
The default is to read all products
in the queue.
.TP
.B -c
Check each product.
Recompute the MD5 checksum of product data and compare it against the
the signature in product description. If the comparison fails,
a message is emitted.

.TP
.B -s
Queue "sanity" check.
Scans entire queue, tallies number of products encountered, and
compares the result with the number of products the queue thinks it should
have.  If the LDM is not running and no product subsets are specified
on the command line, the
results should agree.  This is a simplistic way to determine
whether the queue is corrupted.  Although it may be possible for the
queue to be in a bad state for some other reason, it is expected that
this test will catch most ways in which queue corruption can occur.  Note
that if the LDM is running or if a subset of products is specified on
the command line with the -f or -p options,
the two results will not agree and pqcat will exit with a nonzero
value.  It is intended that pqcat -s be run before starting the LDM in
order to determine whether or not to rebuild the queue before starting.

.SH SIGNALS
.TP
.BR SIGTERM
Normal termination.
.TP
.BR SIGINT
Immediate termination.
.TP
.B SIGUSR1
Write status and product statistics to log output.
.TP
.B SIGUSR2
Cyclically increment the verbosity of the program. Assumming the program was
started without the \fB-v\fP or \fB-x\fP switches, the first \fBSIGUSR2\fP will
make it verbose and \fBLOG_INFO\fP priority messages will appear.
The second will turn on the \fBLOG_DEBUG\fP priority messages as well.
A third will take it back to the normal state.

.SH EXAMPLE

The following invocation will capture into the file /tmp/pq.contents
information about all the products currently in the default product queue:
.RS +4
  pqcat -vl /tmp/pq.contents > /dev/null
.RE
The following example will emit to stderr information about each product as
it is inserted into a product queue in /tmp/ldm.pq, starting at the
current time:
.RS +4
  pqcat -v -q /tmp/ldm.pq -o 0 -i 15 > /dev/null
.RE

.SH FILES
.LP

.SH "SEE ALSO"
.LP
.BR ldmd (1),
.BR pqact(1),
.BR ulog (3),
.BR pq (3),
.BR syslogd (8),
WWW URL \fBhttp://www.unidata.ucar.edu/software/ldm/\fP.

.SH SUPPORT
.LP
If you have problems with this program, then you should first examine the 
LDM email archive for similar problems and how they were solved.
The email archive is available via the following World Wide Web URL:
.sp
.RS
\fBhttp://www.unidata.ucar.edu/software/ldm/\fP
.RE
.sp
If this does not suffice and your site is a member of the Unidata 
program, then send an inquiry via email -- together will all relevant 
information -- to
.sp
.RS
\fBsupport@unidata.ucar.edu\fP
.RE


ldm/src/pqcat/pqcat.c
/*
 *   Copyright 1993, University Corporation for Atmospheric Research
 *   See ../COPYRIGHT file for copying and redistribution conditions.
 */
/* $Id: pqcat.c,v 1.40.10.1.2.4 2008/04/15 16:34:09 steve Exp $ */

/* 
 *  dump a pq
 */

#include <ldmconfig.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <string.h>
#include <rpc/rpc.h>
#include <signal.h>
#include <unistd.h>
#include <errno.h>
#include <assert.h>
#include <regex.h>
#include "ldm.h"
#include "atofeedt.h"
#include "ldmprint.h"
#include "ulog.h"
#include "pq.h"
#include "paths.h"
#include "md5.h"
#include "RegularExpressions.h"

#ifdef NO_ATEXIT
#include "atexit.h"
#endif

/* default "one trip" */
#ifndef DEFAULT_INTERVAL
#define DEFAULT_INTERVAL 0
#endif

#ifndef DEFAULT_FEEDTYPE
#define DEFAULT_FEEDTYPE ANY
#endif

static volatile int done = 0;
static volatile int intr = 0;
static volatile int stats_req = 0;
static int showProdOrigin = 0;

static const char *pqfname = DEFAULT_QUEUE;
static pqueue *pq = NULL;

static int nprods = 0;

static MD5_CTX *md5ctxp = NULL;

static void
dump_stats(void)
{
        unotice("Number of products %d", nprods);
}

/*
 */
/*ARGSUSED*/
static int
writeprod(const prod_info *infop, const void *datap,
                void *xprod, size_t size,  void *notused)
{
        if(ulogIsVerbose())
        {
                if (showProdOrigin)
                {
                        uinfo("%s %s", s_prod_info(NULL, 0, infop, ulogIsDebug()), infop->origin);
                } else
                {
                        uinfo("%s", s_prod_info(NULL, 0, infop, ulogIsDebug()));
                }
        }

        if(md5ctxp != NULL) /* -c option */
        {
                signaturet check;
                MD5Init(md5ctxp);
                MD5Update(md5ctxp, datap, infop->sz);
                MD5Final((unsigned char*)check, md5ctxp);
                if(memcmp(infop->signature, check, sizeof(signaturet)) != 0)
                {
                        char sb[33]; char cb[33];
                        uerror("signature mismatch: %s != %s",
                                 s_signaturet(sb, sizeof(sb), infop->signature),
                                 s_signaturet(cb, sizeof(cb), check));
                }
        }

        if( write(STDOUT_FILENO, datap, infop->sz) !=
                        infop->sz)
        {
                int errnum = errno;
                serror( "data write failed") ;
                return errnum;
        }

        nprods++;

        return 0;
}

static int
tallyProds(const prod_info *infop, const void *datap,
                void *xprod, size_t size,  void *notused)
{
        nprods++;

        return 0;
}

static void
usage(const char *av0) /*  id string */
{
        (void)fprintf(stderr,
                "Usage: %s [options] [outputfile]\n\tOptions:\n", av0);
        (void)fprintf(stderr,
                "\t-v           Verbose, tell me about each product\n");
        (void)fprintf(stderr,
                "\t-l logfile   Log to a file rather than stderr\n");
        (void)fprintf(stderr,
                "\t-f feedtype  Scan for data of type \"feedtype\" (default \"%s\")\n", s_feedtypet(DEFAULT_FEEDTYPE));
        (void)fprintf(stderr,
                "\t-p pattern   Interested in products matching \"pattern\" (default \".*\")\n") ;
        (void)fprintf(stderr,
                "\t-q pqfname   (default \"%s\")\n", DEFAULT_QUEUE);
        (void)fprintf(stderr,
                "\t-o offset    Set the \"from\" time \"offset\" secs before now\n");
        (void)fprintf(stderr,
                "\t             (default \"from\" the beginning of the epoch)\n");
        (void)fprintf(stderr,
                "\t-i interval  Poll queue after \"interval\" secs (default %d)\n",
                DEFAULT_INTERVAL);
        (void)fprintf(stderr,
                "\t             (\"interval\" of 0 means exit at end of queue)\n");
        (void)fprintf(stderr,
                "\t-c           Check, verify MD5 signature\n");
        (void)fprintf(stderr,
                "\t-O           Include product origin in verbose output\n");
        (void)fprintf(stderr,
                "\t             (valid only with -v option)\n");
        (void)fprintf(stderr,
                "\t-s           Check queue for sanity/non-corruption\n");
        (void)fprintf(stderr,
                "Output defaults to standard output\n");
        exit(1);
}


static void
cleanup(void)
{
        unotice("Exiting"); 

        if(!intr)
        {
                if(md5ctxp != NULL)
                        free_MD5_CTX(md5ctxp);  

                if(pq != NULL)  
                        (void)pq_close(pq);
        }

        dump_stats();

        (void) closeulog();
}


static void
signal_handler(int sig)
{
#ifdef SVR3SIGNALS
        /* 
         * Some systems reset handler to SIG_DFL upon entry to handler.
         * In that case, we reregister our handler.
         */
        (void) signal(sig, signal_handler);
#endif
        switch(sig) {
        case SIGINT :
                intr = !0;
                exit(0);
        case SIGTERM :
                done = !0;      
                return;
        case SIGUSR1 :
                stats_req = !0;
                return;
        case SIGUSR2 :
                rollulogpri();
                return;
        }
}


/*
 * register the signal_handler
 */
static void
set_sigactions(void)
{
        struct sigaction sigact;

        sigemptyset(&sigact.sa_mask);
        sigact.sa_flags = 0;

        /* Ignore these */
        sigact.sa_handler = SIG_IGN;
        (void) sigaction(SIGHUP, &sigact, NULL);
        (void) sigaction(SIGPIPE, &sigact, NULL);
        (void) sigaction(SIGALRM, &sigact, NULL);
        (void) sigaction(SIGCHLD, &sigact, NULL);

        /* Handle these */
#ifdef SA_RESTART       /* SVR4, 4.3+ BSD */
        /* usually, restart system calls */
        sigact.sa_flags |= SA_RESTART;
#endif
        sigact.sa_handler = signal_handler;
        (void) sigaction(SIGTERM, &sigact, NULL);
        (void) sigaction(SIGUSR1, &sigact, NULL);
        (void) sigaction(SIGUSR2, &sigact, NULL);

        /* Don't restart after interrupt */
        sigact.sa_flags = 0;
#ifdef SA_INTERRUPT     /* SunOS 4.x */
        sigact.sa_flags |= SA_INTERRUPT;
#endif
        (void) sigaction(SIGINT, &sigact, NULL);
}


int main(int ac, char *av[])
{
        const char *progname = ubasename(av[0]);
        char *logfname;
        prod_class_t clss;
        prod_spec spec;
        int status = 0;
        int interval = DEFAULT_INTERVAL;
        int logoptions = (LOG_CONS|LOG_PID) ;
        int queueSanityCheck = FALSE;

        logfname = "";

        if(isatty(fileno(stderr)))
        {
                /* set interactive defaults */
                logfname = "-" ;
                logoptions = 0 ;
        }

        clss.from = TS_ZERO; /* default dump the whole file */
        clss.to = TS_ENDT;
        clss.psa.psa_len = 1;
        clss.psa.psa_val = &spec;
        spec.feedtype = ANY;
        spec.pattern = ".*";
        
        /*
         * Check the environment for some options.
         * May be overridden by command line switches below.
         */
        {
                const char *ldmpqfname = getenv("LDMPQFNAME");
                if(ldmpqfname != NULL)
                        pqfname = ldmpqfname;
        }

        {
        extern int optind;
        extern int opterr;
        extern char *optarg;
        int ch;
        int logmask = (LOG_MASK(LOG_ERR) | LOG_MASK(LOG_WARNING) |
            LOG_MASK(LOG_NOTICE));
        int fterr;

        opterr = 1;

        while ((ch = getopt(ac, av, "cvxOsl:p:f:q:o:i:")) != EOF)
                switch (ch) {
                case 'c':
                        md5ctxp = new_MD5_CTX();
                        if(md5ctxp == NULL)
                                serror("new_md5_CTX failed");
                        break;
                case 'v':
                        logmask |= LOG_MASK(LOG_INFO);
                        break;
                case 'x':
                        logmask |= LOG_MASK(LOG_DEBUG);
                        break;
                case 'l':
                        logfname = optarg;
                        break;
                case 'p':
                        spec.pattern = optarg;
                        /* compiled below */
                        break;
                case 'f':
                        fterr = strfeedtypet(optarg, &spec.feedtype) ;
                        if(fterr != FEEDTYPE_OK)
                        {
                                fprintf(stderr, "Bad feedtype \"%s\", %s\n",
                                        optarg, strfeederr(fterr)) ;
                                usage(progname);        
                        }
                        break;
                case 'q':
                        pqfname = optarg;
                        break;
                case 'o':
                        (void) set_timestamp(&clss.from);
                        clss.from.tv_sec -= atoi(optarg);
                        break;
                case 'i':
                        interval = atoi(optarg);
                        if(interval == 0 && *optarg != '0')
                        {
                                fprintf(stderr, "%s: invalid interval %s",
                                        progname, optarg);
                                usage(progname);
                        }
                        break;
                case 'O':
                        showProdOrigin = TRUE;
                        break;
                case 's':
                        queueSanityCheck = TRUE;
                        break;
                case '?':
                        usage(progname);
                        break;
                }

        (void) setulogmask(logmask);

        if (re_isPathological(spec.pattern))
        {
                fprintf(stderr, "Adjusting pathological regular-expression: "
                    "\"%s\"\n", spec.pattern);
                re_vetSpec(spec.pattern);
        }
        status = regcomp(&spec.rgx,
                spec.pattern,
                REG_EXTENDED|REG_NOSUB);
        if(status != 0)
        {
                fprintf(stderr, "Bad regular expression \"%s\"\n",
                        spec.pattern);
                usage(av[0]);
        }


        /* last arg, outputfname, is optional */
        if(ac - optind > 0)
        {
                const char *const outputfname = av[optind];
                if(freopen(outputfname, "a+b", stdout) == NULL)
                {
                        status = errno;
                        fprintf(stderr, "%s: Couldn't open \"%s\": %s\n",
                                progname, outputfname, strerror(status));
                }
        }

        }

        /*
         * Set up error logging.
         */
        (void) openulog(progname,
                logoptions, LOG_LDM, logfname);
        unotice("Starting Up (%d)", getpgrp());

        /*
         * register exit handler
         */
        if(atexit(cleanup) != 0)
        {
                serror("atexit");
                exit(1);
        }

        /*
         * set up signal handlers
         */
        set_sigactions();


        /*
         * Open the product que
         */
        status = pq_open(pqfname, PQ_READONLY, &pq);
        if(status)
        {
                if (PQ_CORRUPT == status) {
                    uerror("The product-queue \"%s\" is inconsistent\n",
                            pqfname);
                }
                else {
                    uerror("pq_open failed: %s: %s\n",
                            pqfname, strerror(status));
                }
                exit(1);
        }

        /*
         * Set the cursor position to starting time
         */
        pq_cset(pq, &clss.from );

        while(!done)
        {
                if(stats_req)
                {
                        dump_stats();
                        stats_req = 0;
                }

                if (queueSanityCheck)
                  status = pq_sequence(pq, TV_GT, &clss, tallyProds, 0);
                else
                  status = pq_sequence(pq, TV_GT, &clss, writeprod, 0);

                switch(status) {
                case 0: /* no error */
                        continue; /* N.B., other cases sleep */
                case PQUEUE_END:
                        udebug("End of Queue");
                        break;
                case EAGAIN:
                case EACCES:
                        udebug("Hit a lock");
                        break;
                default:
                        uerror("pq_sequence failed: %s (errno = %d)",
                                strerror(status), status);
                        exit(1);
                        break;
                }

                if(interval == 0)
                        break;
                pq_suspend(interval);
                        
        }

        if (queueSanityCheck)
        {
          size_t queueProdCnt;
          size_t dummy1, dummy2, dummy3, dummy4, dummy5, dummy6, dummy7, dummy9;
          double dummy8;
          status = pq_stats(pq, &queueProdCnt, &dummy1, &dummy2, &dummy3, &dummy4, 
                   &dummy5, &dummy6, &dummy7, &dummy8, &dummy9);
          if (status) {
                uerror("pq_stats failed: %s (errno = %d)",
                       strerror(status), status);
                exit(1);
          }

          if (nprods == queueProdCnt)
          {
            unotice("pqcat queueSanityCheck: Number of products tallied consistent with value in queue");
            exit(0);
          }
          else
          {
              uerror("pqcat queueSanityCheck: Product count doesn't match");
              uerror("products tallied: %d   Value in queue: %d", nprods, queueProdCnt);              exit(1);
          }
        }
        exit(0);
        /*NOTREACHED*/
}
